pkg/gatewayserver/io/io.go

```diff
@@ -214,6 +214,8 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage) error {
 			if location.Source != ttnpb.SOURCE_UNKNOWN {
 				md.Location = &location
 			}
+		} else {
```

Comment:
This happens if the location is public, but the antenna index doesn't match, so maybe

```suggestion
		} else if !c.gateway.LocationPublic {
```